packages/jsii-pacmak/lib/targets/golang/module.ts

```diff
@@ -1,64 +1,254 @@
+  throw new Error(`Type: ${type.name} is not an interface, enum, or class`);
+}
 
-  public constructor(assembly: Assembly, submodule?: JsiiSubmodule) {
-    this.assembly = submodule ?? assembly;
-    assembly.types.forEach((type) => {
-      let t: Enum | Interface | GoClass | undefined;
+function buildModuleTypes(parent: Module, types: readonly Type[]): ModuleTypes {
+  return types.map(
+    (type: Type): ModuleType => {
       if (type.isInterfaceType()) {
-        t = new Interface(type);
+        return new Interface(parent, type);
       } else if (type.isClassType()) {
-        t = new GoClass(type);
+        return new GoClass(parent, type);
       } else if (type.isEnumType()) {
-        t = new Enum(type);
+        return new Enum(parent, type);
       }
+      return unknownType(type);
+    },
+  );
+}
 
-      if (t) {
-        this.types[type.fqn] = t;
-      }
-    });
+function findTypeInTree(module: Module, fqn: string): ModuleType | undefined {
+  const result = module.types.find((t) => t.type.fqn === fqn);
 
-    this.submodules = this.buildSubmodules(assembly, submodule);
+  if (result) {
+    return result;
+  } else if (module.hasSubmodules()) {
+    return module.submodules.reduce((accum: ModuleType | undefined, sm) => {
+      return accum || findTypeInTree(sm, fqn);
+    }, undefined);
   }
 
-  public buildSubmodules(
-    assembly: Assembly,
-    submodule?: JsiiSubmodule,
-  ): Submodule[] {
-    return (submodule?.submodules ?? assembly.submodules).map(
-      (sm) => new Submodule(assembly, sm),
+  return undefined;
+}
+
+const JSII_MODULE_NAME = 'github.com/aws-cdk/jsii/jsii';
+export abstract class ModuleFile {
+  public constructor(public readonly file: string) {}
+  public open(code: CodeMaker): void {
+    code.openFile(this.file);
+  }
+
+  public close(code: CodeMaker): void {
+    code.closeFile(this.file);
+  }
+}
+
+export class RootModule extends ModuleFile implements Module {
+  private readonly assembly: Assembly;
+  public readonly packageName: string;
+  public readonly moduleName: string;
+  public readonly file: string;
+  public readonly filePath: string;
+  public readonly root: RootModule;
+
+  public constructor(assembly: Assembly) {
+    const packageName = assembly.name
+      .replace('@', '')
+      .replace(/[^a-z0-9_.]/gi, '');
+    const filePath = join(...packageName.split('.'));
+    const file = `${filePath}.go`;
+    super(file);
+
+    this.assembly = assembly;
+    this.root = this;
+    this.packageName = packageName;
+    // moduleName == packageName for root;
+    this.moduleName = packageName;
+    this.filePath = filePath;
+    this.file = file;
+  }
+
+  public get types(): ModuleTypes {
+    return buildModuleTypes(this, Object.values(this.assembly.types));
+  }
+
+  public get submodules(): Submodule[] {
+    return buildSubmodules(this, this, this.assembly.submodules);
+  }
+
+  public get dependencies(): Set<Module> {
+    return new Set(
```

Comment:
Why wrap in a `Set` here?

Reply:
dependencies collect all `Module` parents of type references within the module, so there can be multiple type references that come from the same module. So the Set de-dupes those modules.

Reply:
If you get different *instances* of the `Module` type, then you're actually not de-duping anything...

Reply:
yeah I actually moved it to be a `Set<string>` used for the dependency module names. They were in fact unique instances of Module since it was pulling them from dynamic getters.